packages/decoder/lib/ProviderAdapter.ts

```diff
@@ -0,0 +1,167 @@
+
+// lifted from @types/web3
+type Log = {
+    address: string;
+    data: string;
+    topics: string[];
+    logIndex: number;
+    transactionHash: string;
+    transactionIndex: number;
+    blockHash: string;
+    blockNumber: number;
+}
+type PastLogsOptions = {
+  toBlock?: string | number;
+  fromBlock?: string | number;
+  address?: string | string[];
+};
+type SendRequestArgs = {
+  method: string;
+  params: unknown[];
+};
+type EipElevenNinetyThreeProvider = {
+  request: (options: { method: string; params: unknown[]; }) => Promise<any>;
+}
+const stringWhitelist = [
+  "latest",
+  "pending",
+  "genesis",
+  "earliest"
+];
+
+const formatBlockSpecifier = (block: BlockSpecifier): string => {
+  if (typeof block === "string" && stringWhitelist.includes(block)) {
+    // block is one of 'latest', 'pending', 'earliest', or 'genesis'
+    return block === "genesis" ?
+      // convert old web3 input format which uses 'genesis'
+      "earliest" :
+      block;
+  } else if (typeof block === "string" && !isNaN(parseInt(block))) {
+    // block is a string representation of a number
+    if (block.startsWith("0x")) return block;
+    // convert to hex and add '0x' prefix in case block is decimal
+    return `0x${parseInt(block).toString(16)}`;
+  } else if (typeof block === "number") {
+    return `0x${block.toString(16)}`;
+  } else {
+    throw new Error(
+      "The block specified must be a number or one of the strings 'latest'," +
+      "'pending', or 'earliest'."
+    );
+  }
+};
+
+const isElevenNinetyThreeProvider = (provider: Provider | EipElevenNinetyThreeProvider): provider is EipElevenNinetyThreeProvider => {
+  return typeof (provider as EipElevenNinetyThreeProvider).request === "function";
+};
+
+export class ProviderAdapter {
+  public provider: Provider | EipElevenNinetyThreeProvider;
+
+  constructor (provider: Provider | EipElevenNinetyThreeProvider) {
+    this.provider = provider;
+  }
+
+  private async sendRequest ({
+    method,
+    params
+  }: SendRequestArgs): Promise<any> {
+    if (!this.provider) {
+      throw new Error("There is not a valid provider present.")
+    }
+    // check to see if the provider is compliant with eip1193
+    // https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md
+    if (isElevenNinetyThreeProvider(this.provider)) {
+      return (await this.provider.request({ method, params })).result;
+    } else {
+      const sendMethod = promisify(this.provider.send).bind(this.provider);
+      return (await sendMethod({
+        jsonrpc: "2.0",
+        id: new Date().getTime(),
+        method,
+        params
+      })).result;
+    }
+  }
+
+  public async getCode (address: string, block: RegularizedBlockSpecifier) {
+    const blockToFetch = formatBlockSpecifier(block);
+    return await this.sendRequest({
+      method: "eth_getCode",
+      params: [
+        address,
+        blockToFetch
+      ]
+    });
+  }
+
+  public async getBlock (block: BlockSpecifier) {
```

Comment:
Should this maybe be renamed to `getBlockByNumber`? Looks like web3's `getBlock` method is used for both "eth_getBlockByNumber" and "eth_getBlockByHash"

Reply:
Yeah, I think that rename is a good idea

Reply:
It does also accept "latest", "earliest"....etc.